libraries/AP_RangeFinder/AP_RangeFinder_Wasp.cpp

```diff
@@ -0,0 +1,269 @@
+    // @Range 0-255
+    // @User: Advanced
+    AP_GROUPINFO("WSP_MEDF", 2, AP_RangeFinder_Wasp, medf, 4),
+
+    // @Param: WSP_FRQ
+    // @DisplayName: Frequency
+    // @Description: Sets the repetition frequency of the ranging operation in Hertz. Upon entering the desired frequency the system will calculate the nearest frequency that it can handle according to the resolution of internal timers.
+    // @Range 0-10000
+    // @User: Advanced
+    AP_GROUPINFO("WSP_FRQ", 3, AP_RangeFinder_Wasp, frq, 100),
+
+    // @Param: WSP_AVG
+    // @DisplayName: Multi-pulse averages
+    // @Description: Sets the number of pulses to be used in multi-pulse averaging mode. In this mode, a sequence of rapid fire ranges are taken and then averaged to improve the accuracy of the measurement
+    // @Range 0-255
+    // @User: Advanced
+    AP_GROUPINFO("WSP_AVG", 4, AP_RangeFinder_Wasp, avg, 4),
+
+    // @Param: WSP_THR
+    // @DisplayName: Sensitivity threshold
+    // @Description: Sets the system sensitivity. Larger values of THR represent higher sensitivity. The system may limit the maximum value of THR to prevent excessive false alarm rates based on settings made at the factory
+    // @Range 0-255
+    // @User: Advanced
+    AP_GROUPINFO("WSP_THR", 5, AP_RangeFinder_Wasp, thr, -1),
+
+    AP_GROUPEND
+};
+
+
+/* 
+   The constructor also initialises the rangefinder. Note that this
+   constructor is not called until detect() returns true, so we
+   already know that we should setup the rangefinder
+*/
+AP_RangeFinder_Wasp::AP_RangeFinder_Wasp(RangeFinder::RangeFinder_State &_state,
+                                         AP_SerialManager &serial_manager,
+                                         uint8_t serial_instance) :
+    AP_RangeFinder_Backend(_state) {
+    AP_Param::setup_object_defaults(this, var_info);
+
+    uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance);
+    if (uart != nullptr) {
+        uart->begin(115200);
+
+        // register Wasp specific parameters
+        state.var_info = var_info;
+
+        configuration_state = WASP_CFG_PROTOCOL;
+    }
+}
+
+/* 
+   detect if a Lightware rangefinder is connected. We'll detect by
+   trying to take a reading on Serial. If we get a result the sensor is
+   there.
+*/
+bool AP_RangeFinder_Wasp::detect(AP_SerialManager &serial_manager, uint8_t serial_instance) {
+    return serial_manager.find_serial(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance) != nullptr;
+}
+
+// read - return last value measured by sensor
+bool AP_RangeFinder_Wasp::get_reading(uint16_t &reading_cm) {
+    if (uart == nullptr) {
+        return false;
+    }
+
+    // read any available lines from the lidar
+    float sum = 0;
+    uint16_t count = 0;
+    int16_t nbytes = uart->available();
+    while (nbytes-- > 0) {
+        char c = uart->read();
+        if (c == '\n') {
```

Comment:
check for -1 return from uart->read()

Reply:
@tridge What should be done in that case? The API previously responded with nbytes available, but then it can't read one byte, should code try again or stop immediately?

Reply:
@tridge I have to agree with @OXINARF on this. If `uart->read()` can be expected to fail if we know there are bytes available then we have significant issues across the code base in a number of drivers.

Reply:
@WickedShell I had never thought about, but looking at HAL it certainly can be expected to fail - it might be as simple as not being able to get a lock on the read buffer. My question is more what @tridge is expecting that code does because, as you say, I've never seen it handled in other drivers - and we definitely should at least make sure it's a valid (positive) char (which your driver actually kind of does).

Reply:
ok, I'm convinced, lets ignore this one